File: OpenBurnBarMobile/Settings/Search/SettingsSearchEngine.swift
import Foundation

// MARK: - Settings Search Engine (iOS)

/// Mirrors the macOS engine — weighted, AND-semantic, diacritic-folded
/// substring matching across `title` / `keywords` / `subtitle` / `helpText`.
///
/// The implementation is intentionally duplicated rather than shared, per
/// the cross-platform plan: keep platforms independent so neither has to
/// move in lockstep with the other for incidental changes.
enum SettingsSearchEngine {

    /// Weight applied to a hit in the matching field.
    enum Weight {
        static let title = 3
        static let keyword = 2
        static let subtitle = 2
        static let helpText = 1
    }

    static let defaultResultLimit = 25

    /// Returns ranked matches for `query` across `items`.
    static func search(
        _ query: String,
        in items: [SettingsItem],
        limit: Int = defaultResultLimit
    ) -> [SettingsItem] {
        let tokens = tokenize(query)
        guard !tokens.isEmpty else { return [] }

        let scored: [(item: SettingsItem, score: Int)] = items.compactMap { item in
            guard let score = score(item: item, tokens: tokens) else { return nil }
            return (item, score)
        }

        let ranked = scored.sorted { lhs, rhs in
            if lhs.score != rhs.score { return lhs.score > rhs.score }
            return foldedTitle(lhs.item) < foldedTitle(rhs.item)
        }

        return Array(ranked.prefix(limit).map(\.item))
    }

    // MARK: - Internals

    static func score(item: SettingsItem, tokens: [String]) -> Int? {
        let title = foldedTitle(item)
        let keywords = item.keywords.map(Self.fold)
        let subtitle = item.subtitle.map(Self.fold) ?? ""
        let helpText = item.helpText.map(Self.fold) ?? ""

        var total = 0
        for token in tokens {
            var tokenScore = 0
            if title.contains(token) { tokenScore += Weight.title }
            if keywords.contains(where: { $0.contains(token) }) {
                tokenScore += Weight.keyword
            }
            if !subtitle.isEmpty, subtitle.contains(token) {
                tokenScore += Weight.subtitle
            }
            if !helpText.isEmpty, helpText.contains(token) {
                tokenScore += Weight.helpText
            }
            if tokenScore == 0 { return nil }
            total += tokenScore
        }
        return total
    }

    static func fold(_ input: String) -> String {
        input.folding(options: [.diacriticInsensitive, .caseInsensitive],
                      locale: .current)
    }

    static func tokenize(_ query: String) -> [String] {
        fold(query)
            .split(whereSeparator: { $0.isWhitespace })
            .map(String.init)
            .filter { !$0.isEmpty }
    }

    private static func foldedTitle(_ item: SettingsItem) -> String {
        fold(item.title)
    }
}

File: android/app/src/main/java/com/openburnbar/ui/settings/SettingsSearchEngine.kt
package com.openburnbar.ui.settings

import java.text.Normalizer
import java.util.Locale

/**
 * Settings search ranking engine — Android port of the macOS/iOS version.
 *
 * Same semantics:
 * - Weighted hits: title=3, keywords=2, subtitle=2, helpText=1.
 * - AND token semantics: every token must hit somewhere on the row.
 * - Diacritic-folded, case-insensitive substring match.
 * - Tie-break by lowercase title ascending. Capped at [DEFAULT_RESULT_LIMIT].
 */
object SettingsSearchEngine {

    const val WEIGHT_TITLE = 3
    const val WEIGHT_KEYWORD = 2
    const val WEIGHT_SUBTITLE = 2
    const val WEIGHT_HELP_TEXT = 1

    const val DEFAULT_RESULT_LIMIT = 25

    /** Returns the items ranked against [query]. Empty/whitespace yields []. */
    fun search(
        query: String,
        items: List<SettingsItem>,
        limit: Int = DEFAULT_RESULT_LIMIT,
    ): List<SettingsItem> {
        val tokens = tokenize(query)
        if (tokens.isEmpty()) return emptyList()

        val scored = items.mapNotNull { item ->
            score(item, tokens)?.let { item to it }
        }

        return scored
            .sortedWith(compareByDescending<Pair<SettingsItem, Int>> { it.second }
                .thenBy { foldedTitle(it.first) })
            .take(limit)
            .map { it.first }
    }

    internal fun score(item: SettingsItem, tokens: List<String>): Int? {
        val title = foldedTitle(item)
        val keywords = item.keywords.map { fold(it) }
        val subtitle = item.subtitle?.let { fold(it) }.orEmpty()
        val helpText = item.helpText?.let { fold(it) }.orEmpty()

        var total = 0
        for (token in tokens) {
            var tokenScore = 0
            if (title.contains(token)) tokenScore += WEIGHT_TITLE
            if (keywords.any { it.contains(token) }) tokenScore += WEIGHT_KEYWORD
            if (subtitle.isNotEmpty() && subtitle.contains(token)) tokenScore += WEIGHT_SUBTITLE
            if (helpText.isNotEmpty() && helpText.contains(token)) tokenScore += WEIGHT_HELP_TEXT
            if (tokenScore == 0) return null
            total += tokenScore
        }
        return total
    }

    /** Lowercase + remove combining diacritics for case-insensitive match. */
    internal fun fold(input: String): String {
        val normalized = Normalizer.normalize(input, Normalizer.Form.NFD)
        // Strip combining marks (e.g. accents) so "café" matches "cafe".
        val stripped = normalized.replace(Regex("\\p{InCombiningDiacriticalMarks}+"), "")
        return stripped.lowercase(Locale.ROOT)
    }

    internal fun tokenize(query: String): List<String> =
        fold(query)
            .split(Regex("\\s+"))
            .filter { it.isNotEmpty() }

    private fun foldedTitle(item: SettingsItem): String = fold(item.title)
}
